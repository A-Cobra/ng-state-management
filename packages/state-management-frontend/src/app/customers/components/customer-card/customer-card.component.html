<clapp-card>
  <clapp-card-content>
    <div class="customer">
      <clapp-image-display type="avatar" size="medium"></clapp-image-display>
      <div class="customer__information">
        <h3>{{ customer.name }} {{ customer.lastName }}</h3>
        <p>@{{ customer.username }}</p>
      </div>
    </div>
  </clapp-card-content>
</clapp-card>
