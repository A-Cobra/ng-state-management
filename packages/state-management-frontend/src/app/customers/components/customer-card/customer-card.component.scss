.customer {
  display: flex;
  gap: 1rem;

  &__information {
    display: flex;
    flex-direction: column;
    justify-content: center;
    gap: 0.3rem;
  }
}
