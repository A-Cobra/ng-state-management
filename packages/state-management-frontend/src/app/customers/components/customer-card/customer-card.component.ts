import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { CustomerInterface } from '@state-management-app/types';

@Component({
  selector: 'app-customer-card',
  templateUrl: './customer-card.component.html',
  styleUrls: ['./customer-card.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class CustomerCardComponent {
  @Input() customer: CustomerInterface;
}
