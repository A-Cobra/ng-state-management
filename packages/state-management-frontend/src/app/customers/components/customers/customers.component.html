<div class="action-buttons">
  <!-- TODO: Redirect to previous page -->
  <clapp-button color="secondary" plusIcon="ri-arrow-left-s-line">
    Go back
  </clapp-button>
  <clapp-search
    [formControl]="searchCustomerControl"
    placeholder="Search branch by name"
  />
  <div class="spacer"></div>
</div>

<ng-container *ngIf="customers$ | async as customers">
  <ng-container *ngIf="totalRecords; else noResults">
    <div class="customers">
      <state-management-app-customer-card
        *ngFor="let customer of customers"
        [customer]="customer"
      />
    </div>

    <clapp-pagination
      *ngIf="totalRecords > pageSize"
      [totalRecords]="totalRecords"
      [pageSize]="pageSize"
      (pageChange)="changePage($event)"
    />
  </ng-container>

  <ng-template #noResults>
    <clapp-no-results-found customText="Sorry, we couldn't find any branch." />
  </ng-template>
</ng-container>
