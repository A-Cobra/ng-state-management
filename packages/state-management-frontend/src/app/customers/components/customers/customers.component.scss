:host {
  display: flex;
  flex-direction: column;
  gap: 3rem;

  clapp-button,
  .spacer {
    width: 100%;
    max-width: 14rem;

    // TODO: Refactor when mixins for media queries are implemented
    @media (max-width: 480px) {
      max-width: unset;

      &:is(.spacer) {
        display: none;
      }
    }

    ::ng-deep {
      button {
        width: 100%;
      }
    }
  }
}

.action-buttons {
  display: flex;
  justify-content: space-between;
  gap: 1.5rem;

  // TODO: Refactor when mixins for media queries are implemented
  @media (max-width: 480px) {
    flex-direction: column;
  }

  clapp-search {
    width: 100%;
    max-width: 30rem;
  }
}

.customers {
  display: grid;
  grid-template-columns: repeat(auto-fill, 280px);
  grid-gap: 1.5rem;
  justify-content: center;
}
