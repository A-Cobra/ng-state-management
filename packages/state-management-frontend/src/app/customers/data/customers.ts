import { Customer } from '../models/customer.model';

export const CUSTOMERS: Customer[] = [
  {
    id: '1',
    name: 'John',
    lastName: 'Doe',
    username: 'johndoe',
    phoneNumber: '123-456-7890',
    email: 'johndoe@example.com',
    paymentMethod: 'Visa',
  },
  {
    id: '2',
    name: 'Jane',
    lastName: 'Doe',
    username: 'janedoe',
    phoneNumber: '234-567-8901',
    email: 'janedoe@example.com',
    paymentMethod: 'Mastercard',
  },
  {
    id: '3',
    name: 'John',
    lastName: 'Smith',
    username: 'johnsmith',
    phoneNumber: '345-678-9012',
    email: 'johnsmith@example.com',
    paymentMethod: 'Visa',
  },
  {
    id: '4',
    name: 'Jane',
    lastName: 'Smith',
    username: 'janesmith',
    phoneNumber: '456-789-0123',
    email: 'janesmith@example.com',
    paymentMethod: 'Mastercard',
  },
  {
    id: '5',
    name: 'Mark',
    lastName: 'Smith',
    username: 'marksmith',
    phoneNumber: '567-890-1234',
    email: 'marksmith@example.com',
    paymentMethod: 'Visa',
  },
  {
    id: '6',
    name: 'Steve',
    lastName: 'Jobs',
    username: 'stevejobs',
    phoneNumber: '678-901-2345',
    email: 'stevejobs@example.com',
    paymentMethod: 'Mastercard',
  },
  {
    id: '7',
    name: 'John',
    lastName: 'Doe',
    username: 'johndoe',
    phoneNumber: '789-012-3456',
    email: 'johndoe@example.com',
    paymentMethod: 'Visa',
  },
  {
    id: '8',
    name: 'Jane',
    lastName: 'Doe',
    username: 'janedoe',
    phoneNumber: '890-123-4567',
    email: 'janedoe@example.com',
    paymentMethod: 'Visa',
  },
  {
    id: '9',
    name: 'John',
    lastName: 'Smith',
    username: 'johnsmith',
    phoneNumber: '901-234-5678',
    email: 'johnsmith@example.com',
    paymentMethod: 'Visa',
  },
  {
    id: '10',
    name: 'Jane',
    lastName: 'Smith',
    username: 'janesmith',
    phoneNumber: '012-345-6789',
    email: 'janesmith@example.com',
    paymentMethod: 'Visa',
  },
  {
    id: '11',
    name: 'Mark',
    lastName: 'Smith',
    username: 'marksmith',
    phoneNumber: '123-456-7890',
    email: 'marksmith@example.com',
    paymentMethod: 'Visa',
  },
  {
    id: '12',
    name: 'Steve',
    lastName: 'Jobs',
    username: 'stevejobs',
    phoneNumber: '234-567-8901',
    email: 'stevejobs@example.com',
    paymentMethod: 'Mastercard',
  },
  {
    id: '13',
    name: 'Mark',
    lastName: 'Smith',
    username: 'marksmith',
    phoneNumber: '345-678-9012',
    email: 'marksmith@example.com',
    paymentMethod: 'Mastercard',
  },
  {
    id: '14',
    name: 'Steve',
    lastName: 'Jobs',
    username: 'stevejobs',
    phoneNumber: '456-789-0123',
    email: 'stevejobs@example.com',
    paymentMethod: 'Mastercard',
  },
];
