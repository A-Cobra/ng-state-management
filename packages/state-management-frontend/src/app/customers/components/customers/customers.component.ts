import { Component, inject } from '@angular/core';
import { FormControl } from '@angular/forms';

import { Customer } from '../../models/customer.model';
import { CustomersService } from '../../services/customers.service';

import {
  debounceTime,
  distinctUntilChanged,
  map,
  Observable,
  Subject,
  takeUntil,
  tap,
} from 'rxjs';

import { Pagination } from '@clapp1/clapp-angular/lib/pagination/interfaces/pagination.interface';

@Component({
  selector: 'state-management-app-customers',
  templateUrl: './customers.component.html',
  styleUrls: ['./customers.component.scss'],
})
export class CustomersComponent {
  readonly #customersService = inject(CustomersService);
  readonly #unsubscribe$ = new Subject<void>();
  searchCustomerControl = new FormControl('', { nonNullable: true });
  customers$: Observable<Customer[]>;
  totalRecords = 0;
  pageSize = 12;

  public ngOnInit(): void {
    this.customers$ = this.#customersService
      .getCustomers(1, this.pageSize)
      .pipe(
        tap((response) => {
          this.totalRecords = response.meta.total;
          this.pageSize = response.meta.pageSize;
        }),
        map((response) => response.data)
      );
    this.searchCustomerControl.valueChanges
      .pipe(
        takeUntil(this.#unsubscribe$),
        debounceTime(700),
        distinctUntilChanged()
      )
      .subscribe((customerName) => {
        this.searchBranch(customerName);
      });
  }

  public searchBranch(customerName: string): void {
    // TODO: implement search by branch name functionality when BE will be ready.
    this.customers$ = this.#customersService
      .getCustomers(1, this.pageSize, customerName)
      .pipe(
        tap((response) => {
          this.totalRecords = response.meta.total;
        }),
        map((response) => response.data)
      );
  }

  public changePage(event: Pagination): void {
    // TODO: Verify implementation when BE will be ready.
    this.customers$ = this.#customersService
      .getCustomers(
        event.currentPage,
        this.pageSize,
        this.searchCustomerControl.value
      )
      .pipe(map((response) => response.data));
  }

  public ngOnDestroy(): void {
    this.#unsubscribe$.next();
    this.#unsubscribe$.complete();
  }
}
