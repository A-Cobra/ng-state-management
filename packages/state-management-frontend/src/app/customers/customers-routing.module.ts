import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';

import { CustomerDetailsComponent } from './components/customer-details/customer-details.component';
import { CustomersComponent } from './components/customers/customers.component';
import { CustomerOrderSelectAddressComponent } from './components/customer-order-select-address/customer-order-select-address.component';

const routes: Routes = [
  {
    path: '',
    component: CustomersComponent,
  },
  {
    path: ':customerId',
    component: CustomerDetailsComponent,
  },
  {
    path: 'order',
    data: { title: 'Order' },
    children: [
      {
        path: 'select-address',
        component: CustomerOrderSelectAddressComponent,
      },
      {
        path: 'create-address',
        component: CustomerOrderSelectAddressComponent,
      },
    ],
  },
];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule],
})
export class CustomersRoutingModule {}
