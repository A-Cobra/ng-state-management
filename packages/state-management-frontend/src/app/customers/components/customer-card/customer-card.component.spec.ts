import { ComponentFixture, TestBed } from '@angular/core/testing';
import { CustomerCardComponent } from './customer-card.component';
import { MOCK_CUSTOMER } from '../../test/customers.mocks';

import {
  ClappCardModule,
  ClappImageDisplayModule,
} from '@clapp1/clapp-angular';

describe('CustomerCardComponent', () => {
  let component: CustomerCardComponent;
  let fixture: ComponentFixture<CustomerCardComponent>;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [CustomerCardComponent],
      imports: [ClappImageDisplayModule, ClappCardModule],
    }).compileComponents();

    fixture = TestBed.createComponent(CustomerCardComponent);
    component = fixture.componentInstance;
    component.customer = MOCK_CUSTOMER;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should render customer information', () => {
    expect(fixture).toMatchSnapshot();
  });
});
