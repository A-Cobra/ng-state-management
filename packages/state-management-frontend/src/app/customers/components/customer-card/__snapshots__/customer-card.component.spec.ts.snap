// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomerCardComponent should render customer information 1`] = `
<app-customer-card
  customer={[Function Object]}
>
  <clapp-card>
    <div
      class="clapp-card-wrapper"
    >
      <clapp-card-content>
        <div
          class="clapp-card-content-wrapper"
        >
          <div
            class="customer"
          >
            <clapp-image-display
              size="medium"
              type="avatar"
            >
              <img
                alt="placeholder"
                class="avatar medium"
                src="assets/clapp/placeholder-avatar.png"
              />
            </clapp-image-display>
            <div
              class="customer__information"
            >
              <h3>
                John Doe
              </h3>
              <p>
                @johndoe
              </p>
            </div>
          </div>
        </div>
      </clapp-card-content>
    </div>
  </clapp-card>
</app-customer-card>
`;
